In [ ]:
import numpy as np

from sklearn.metrics import (
    confusion_matrix,
    classification_report,
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    matthews_corrcoef,
    balanced_accuracy_score
)

In [ ]:
TP = 50
FN = 40
FP = 32
TN = 378

print("TP =", TP)
print("FN =", FN)
print("FP =", FP)
print("TN =", TN)

TP = 50
FN = 40
FP = 32
TN = 378


In [ ]:
manual_matrix = np.array([
    [TP, FN],
    [FP, TN]
])

print(manual_matrix)

[[ 50  40]
 [ 32 378]]


In [ ]:
accuracy = (TP + TN) / (TP + TN + FP + FN)
precision = TP / (TP + FP)
recall = TP / (TP + FN)
specificity = TN / (TN + FP)
f1 = 2 * precision * recall / (precision + recall)

print(f"Accuracy:    {accuracy:.3f}")
print(f"Precision:   {precision:.3f}")
print(f"Recall:      {recall:.3f}")
print(f"Specificity: {specificity:.3f}")
print(f"F1-score:    {f1:.3f}")

Accuracy:    0.856
Precision:   0.610
Recall:      0.556
Specificity: 0.922
F1-score:    0.581


In [ ]:
print(f"Accuracy = ({TP} + {TN}) / ({TP} + {TN} + {FP} + {FN}) = {accuracy:.3f}")
print(f"Precision = {TP} / ({TP} + {FP}) = {precision:.3f}")
print(f"Recall = {TP} / ({TP} + {FN}) = {recall:.3f}")
print(f"Specificity = {TN} / ({TN} + {FP}) = {specificity:.3f}")
print(f"F1 = 2 * {precision:.3f} * {recall:.3f} / ({precision:.3f} + {recall:.3f}) = {f1:.3f}")

Accuracy = (50 + 378) / (50 + 378 + 32 + 40) = 0.856
Precision = 50 / (50 + 32) = 0.610
Recall = 50 / (50 + 40) = 0.556
Specificity = 378 / (378 + 32) = 0.922
F1 = 2 * 0.610 * 0.556 / (0.610 + 0.556) = 0.581


In [ ]:
total = TP + FN + FP + TN

actual_no = FP + TN
always_no_accuracy = actual_no / total

print(f"Always-No Accuracy: {always_no_accuracy:.3f}")
print(f"Our Model Accuracy:  {accuracy:.3f}")

Always-No Accuracy: 0.820
Our Model Accuracy:  0.856


The model accuracy is 0.856, while a model that always predicts that the customer will stay has an accuracy of 0.820.

Therefore, accuracy alone can be misleading because most customers stay. Precision, recall, and F1-score should also be considered.

A false negative is more costly in this scenario because the company fails to identify a customer who is actually going to leave.

A false positive may only lead to an unnecessary retention offer, while a false negative may result in losing the customer.

Therefore, recall should be maximized.

In [ ]:
TP_new = TP + 25
FN_new = FN - 25

FP_new = FP + 85
TN_new = TN - 85

print("New TP =", TP_new)
print("New FN =", FN_new)
print("New FP =", FP_new)
print("New TN =", TN_new)

New TP = 75
New FN = 15
New FP = 117
New TN = 293


In [ ]:
precision_new = TP_new / (TP_new + FP_new)
recall_new = TP_new / (TP_new + FN_new)
f1_new = 2 * precision_new * recall_new / (precision_new + recall_new)

print(f"New Precision: {precision_new:.3f}")
print(f"New Recall:    {recall_new:.3f}")
print(f"New F1-score:  {f1_new:.3f}")

New Precision: 0.391
New Recall:    0.833
New F1-score:  0.532


In [ ]:
print("Original vs New")
print(f"Precision: {precision:.3f} -> {precision_new:.3f}")
print(f"Recall:    {recall:.3f} -> {recall_new:.3f}")
print(f"F1-score:  {f1:.3f} -> {f1_new:.3f}")

Original vs New
Precision: 0.610 -> 0.391
Recall:    0.556 -> 0.833
F1-score:  0.581 -> 0.532


In [ ]:
y_true = np.array(
    [1] * TP +
    [1] * FN +
    [0] * FP +
    [0] * TN
)

y_pred = np.array(
    [1] * TP +
    [0] * FN +
    [1] * FP +
    [0] * TN
)

print("Number of samples:", len(y_true))

Number of samples: 500


In [ ]:
cm = confusion_matrix(y_true, y_pred)

print(cm)

[[378  32]
 [ 40  50]]


In [ ]:
print(classification_report(y_true, y_pred, digits=3))

              precision    recall  f1-score   support

           0      0.904     0.922     0.913       410
           1      0.610     0.556     0.581        90

    accuracy                          0.856       500
   macro avg      0.757     0.739     0.747       500
weighted avg      0.851     0.856     0.853       500



In [ ]:
sk_accuracy = accuracy_score(y_true, y_pred)
sk_precision = precision_score(y_true, y_pred)
sk_recall = recall_score(y_true, y_pred)
sk_f1 = f1_score(y_true, y_pred)

print(f"Accuracy:  {sk_accuracy:.3f}")
print(f"Precision: {sk_precision:.3f}")
print(f"Recall:    {sk_recall:.3f}")
print(f"F1-score:  {sk_f1:.3f}")

Accuracy:  0.856
Precision: 0.610
Recall:    0.556
F1-score:  0.581


In [ ]:
balanced_accuracy_manual = (recall + specificity) / 2

print(f"Balanced Accuracy (manual): {balanced_accuracy_manual:.3f}")

Balanced Accuracy (manual): 0.739


In [ ]:
mcc_manual = (
    (TP * TN) - (FP * FN)
) / np.sqrt(
    (TP + FP) *
    (TP + FN) *
    (TN + FP) *
    (TN + FN)
)

print(f"MCC (manual): {mcc_manual:.3f}")

MCC (manual): 0.495


In [ ]:
mcc_sklearn = matthews_corrcoef(y_true, y_pred)
balanced_accuracy_sklearn = balanced_accuracy_score(y_true, y_pred)

print(f"MCC (sklearn): {mcc_sklearn:.3f}")
print(f"Balanced Accuracy (sklearn): {balanced_accuracy_sklearn:.3f}")

MCC (sklearn): 0.495
Balanced Accuracy (sklearn): 0.739


In [ ]:
print("Manual vs sklearn")
print(f"MCC:               {mcc_manual:.3f} vs {mcc_sklearn:.3f}")
print(f"Balanced Accuracy: {balanced_accuracy_manual:.3f} vs {balanced_accuracy_sklearn:.3f}")

Manual vs sklearn
MCC:               0.495 vs 0.495
Balanced Accuracy: 0.739 vs 0.739


MCC is the fairest overall metric for this model. Accuracy is relatively high at 0.856, but it can be misleading because most customers stay. F1-score focuses on the positive class and is useful for churn detection, but it does not consider true negatives. MCC considers TP, TN, FP, and FN at the same time, so it gives a more balanced evaluation of the model.